etcd-defrag/etcddefrag_controller.go

```diff
@@ -0,0 +1,294 @@
+		}
+	}
+
+	memberHealth, err := r.etcdClient.MemberHealth(ctx)
+	if err != nil {
+		return err
+	}
+
+	if !etcdcli.IsClusterHealthy(memberHealth) {
+		r.log.Error(err, "Cluster is unhealthy", "status", memberHealth.Status())
+		return fmt.Errorf("cluster is unhealthy, status: %s", memberHealth.Status())
+	}
+
+	// filter out learner members since they don't support the defragment API call
+	var etcdMembers []*etcdserverpb.Member
+	for _, m := range members {
+		if !m.IsLearner {
+			etcdMembers = append(etcdMembers, m)
+		}
+	}
+
+	var endpointStatus []*clientv3.StatusResponse
+	var leader *clientv3.StatusResponse
+	for _, member := range etcdMembers {
+		if len(member.ClientURLs) == 0 {
+			// skip unstarted member
+			continue
+		}
+		status, err := r.etcdClient.Status(ctx, member.ClientURLs[0])
+		if err != nil {
+			return err
+		}
+		if leader == nil && status.Leader == member.ID {
+			leader = status
+			continue
+		}
+		endpointStatus = append(endpointStatus, status)
+	}
+
+	// Leader last if possible.
+	if leader != nil {
+		r.log.Info("Appending leader last", "ID", leader.Header.MemberId)
+		endpointStatus = append(endpointStatus, leader)
+	}
+
+	successfulDefrags := 0
+	var errs []error
+	for _, status := range endpointStatus {
+		member, err := getMemberFromStatus(etcdMembers, status)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		// Check each member's status which includes the db size on disk "DbSize" and the db size in use "DbSizeInUse"
+		// compare the % difference and if that difference is over the max diff threshold and also above the minimum
+		// db size we defrag the members state file. In the case where this command only partially completed controller
+		// can clean that up on the next sync. Having the db sizes slightly different is not a problem in itself.
+		if r.isEndpointBackendFragmented(member, status) {
+			fragmentedPercentage := checkFragmentationPercentage(status.DbSize, status.DbSizeInUse)
+			r.log.Info("Member is over defrag threshold", "name", member.Name, "URL", member.ClientURLs[0], "fragmentation percentage", fragmentedPercentage, "DBSize on disk", status.DbSize, "DBSize in use", status.DbSizeInUse, "leader", status.Leader)
+			if _, err := r.etcdClient.Defragment(ctx, member); err != nil {
+				// Defrag can timeout if defragmentation takes longer than etcdcli.DefragDialTimeout.
+				r.log.Error(err, "DefragController Defragment Failed", "member", member.Name, "ID", member.ID)
+				errMsg := fmt.Sprintf("failed defrag on member: %s, memberID: %x: %v", member.Name, member.ID, err)
+				errs = append(errs, fmt.Errorf(errMsg))
+				continue
+			}
+
+			r.log.Info("DefragController Defragment Success", "member", member.Name, "ID", member.ID)
+			successfulDefrags++
+
+			// Give cluster time to recover before we move to the next member.
+			if err := wait.Poll(
```

Comment:
nit: this seems deprecated in favour of PollUntilContextTimeout

Reply:
Hmm, this is from cluster-etcd-operator so I'm not inclined to mess with it so much.